o1vm/src/pickles/proof.rs

```diff
@@ -10,11 +11,87 @@ pub struct WitnessColumns<G, S> {
     pub selector: S,
 }
 
-pub struct ProofInputs<G: KimchiCurve> {
-    pub evaluations: WitnessColumns<Vec<G::ScalarField>, Vec<G::ScalarField>>,
+pub struct ProofInputs<F: Field> {
+    pub evaluations: WitnessColumns<Vec<F>, Vec<F>>,
+}
+
+pub struct NotInversedProofInputs<F: Field> {
+    pub scratch: [Vec<ToInverseOrNot<F>>; crate::interpreters::mips::witness::SCRATCH_SIZE],
+    pub instruction_counter: Vec<F>,
+    pub error: Vec<F>,
+    pub selector: Vec<F>,
+}
+
+impl<F: ark_ff::Field> From<NotInversedProofInputs<F>> for ProofInputs<F> {
+    // This function collects the values marked as to inverse from
+    // the proof input, performs a batch inversion on them,
+    // and returns a usable proof input
+    fn from(not_inversed: NotInversedProofInputs<F>) -> ProofInputs<F> {
+        // Initialising the result
+        let mut res = ProofInputs::new(not_inversed.scratch[1].len());
+        res.evaluations.error = not_inversed.error;
+        res.evaluations.instruction_counter = not_inversed.instruction_counter;
+        res.evaluations.selector = not_inversed.selector;
+
+        // Collecting the values to inverse
+        let to_inverse_map: &mut HashMap<_, F> = &mut HashMap::new();
```

Comment:
Okay, I kind of like the idea of having a hash map to avoid repetitions of inverses computations, but at the same time I wonder if the memory overhead is worth the small chance of repeating field elements to be inverted (unless it's very obvious that we always invert the same set of "trivial" values, and by that I mean much smaller sample size than the actual field domain).

Reply:
The chance of inverting the same element depends heavily on the use case, so it's hard to say.
Honnestly, I was using a hash map to have a key value mapping more than anything else.
As for the memory overhead, which one are you referring to ?

Reply:
I was referring to the actual hash map structure. Not sure how large it can take in memory.